pyphi/tpm.py

```diff
@@ -50,11 +50,35 @@ def condition_tpm_nb(
     tpm, fixed_nodes, state, num_states_per_node=None, node_labels=None
 ):
     df = tpm2df(tpm, num_states_per_node, node_labels)
-    for c in fixed_nodes:
-        df = df.iloc[df.index.get_level_values(c) == state[node_labels.index(c)]]
-    free_nodes = list(set(node_labels) - set(fixed_nodes))
+    num_states_per_node = list(num_states_per_node)
+
+    for c in reversed(fixed_nodes):
+        bool_array = [value == state[c] for value in df.index.get_level_values(c)]
+        df = df.iloc[bool_array]
+        del num_states_per_node[c]
+    
+    fixed_node_labels = [node_labels[fixed_node] for fixed_node in fixed_nodes]
+
+    # TODO This, and probably more of the method, might be improvable
+    # if working with just the NodeLabel class instead of lists here
+    # However, subtracting sets was giving poor ordering, causing errors
+    # downstream, so for now subtracting lists works.
+    def subtract_nodes(a, b):
+        a, b = list(a), list(b)
+        for i in b:
+            if i in a:
+                a.remove(i)
+        return a
+
+    free_nodes = subtract_nodes(node_labels, fixed_node_labels)
+
+    node_indices = tuple(range(len(free_nodes)))
+
     tpmdf = df.groupby(free_nodes, axis="columns").sum()
-    return tpmdf, tpmdf.values
+    for node in fixed_nodes:
+        tpmdf = tpmdf.droplevel(level=node_labels[node], axis=0)
+
+    return tpmdf, tpmdf.values, tuple(num_states_per_node), NodeLabels(free_nodes, node_indices, True), node_indices
 
 
 def condition_tpm(tpm, fixed_nodes, state):
```

Comment:
This function is also called in the binary case, so it needs to work for both. For now you can just make two methods, one identical to the first that is called in the binary case, and another with your implementation that's called in the nonbinary case.

You can run the tests for the `tpm` module using `pytest test/test_tpm.py` to check that the binary case is working.

Reply:
Oops, this was meant to refer to `infer_edge`